### ISSUE TITLE
Prism 3.2.4 no longer returns `access-control-allow-headers` for OPTIONS

### ISSUE THREAD

**Issue by klarose** [stoplightio/prism#956]
**Describe the bug**
We recently upgraded to prism 3.2.4. Our tests started failing due to cors errors. The failing tests send authorization headers to the API. The chromium browser interacting with the prism server would fail to send the request due to cors failing. We have narrowed down this change in behaviour to 3.2.4 -- presumably because the new http server's handler for OPTIONS is not setting the `access-control-allow-headers header`.

**To Reproduce**

1. Using the example open-api document `examples/callbacks/payment-service.oas3.yaml`
2. Run prism 3.2.4  `./prism-3.2.4 mock -p 4010 examples/callbacks/payment-service.oas3.yaml`
3. Then run an OPTIONS curl: `curl -v -XOPTIONS -H @/tmp/headers.txt http://127.0.0.1:4010/invoices/123/subscribe`
4. Note how access-control-allow-headers is not in the result.

E.g.

```
curl -v -XOPTIONS -H @/tmp/headers.txt http://127.0.0.1:4010/invoices/123/subscribe*   Trying 127.0.0.1...
* TCP_NODELAY set
* Connected to 127.0.0.1 (127.0.0.1) port 4010 (#0)
> OPTIONS /invoices/123/subscribe HTTP/1.1
> Host: 127.0.0.1:4010
> Accept: */*
> Connection: keep-alive
> Access-Control-Request-Method: GET
> Sec-Fetch-Mode: cors
> Access-Control-Request-Headers: authorization
> Origin: http://localhost:9876
> Sec-Fetch-Site: cross-site
> Referer: http://localhost:9876/context.html
> User-Agent: Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/78.0.3882.0 Safari/537.36
> Accept-Encoding: gzip, deflate, br
> Accept-Language: en-GB,en-US;q=0.9,en;q=0.8
> 
< HTTP/1.1 204 No Content
< Access-Control-Allow-Origin: http://localhost:9876
< Access-Control-Allow-Credentials: true
< Access-Control-Allow-Methods: GET,DELETE,HEAD,PATCH,POST,PUT
< Date: Wed, 29 Jan 2020 16:45:08 GMT
< Connection: keep-alive
< 
* Connection #0 to host 127.0.0.1 left intact

```

Some notes:

1. `prism-3.2.4` was downloaded from the prism releases: `wget https://github.com/stoplightio/prism/releases/download/v3.2.4/prism-cli-linux -O prism-3.2.4`
2. You can generate /tmp/headers.txt as follows:
```
cat <<EOF >/tmp/headers.txt
Connection: keep-alive
Access-Control-Request-Method: GET
Sec-Fetch-Mode: cors
Access-Control-Request-Headers: authorization
Origin: http://localhost:9876
Sec-Fetch-Site: cross-site
Referer: http://localhost:9876/context.html
User-Agent: Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/78.0.3882.0 Safari/537.36
Accept-Encoding: gzip, deflate, br
Accept-Language: en-GB,en-US;q=0.9,en;q=0.8
EOF
```

**Expected behavior**
I expect my browser to not fail the CORS check. 3.2.3 worked. Here are the results from the OPTIONS check for it:

```
curl -v -XOPTIONS -H @/tmp/headers.txt http://127.0.0.1:4010/invoices/123/subscribe*   Trying 127.0.0.1...
* TCP_NODELAY set
* Connected to 127.0.0.1 (127.0.0.1) port 4010 (#0)
> OPTIONS /invoices/123/subscribe HTTP/1.1
> Host: 127.0.0.1:4010
> Accept: */*
> Connection: keep-alive
> Access-Control-Request-Method: GET
> Sec-Fetch-Mode: cors
> Access-Control-Request-Headers: authorization
> Origin: http://localhost:9876
> Sec-Fetch-Site: cross-site
> Referer: http://localhost:9876/context.html
> User-Agent: Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/78.0.3882.0 Safari/537.36
> Accept-Encoding: gzip, deflate, br
> Accept-Language: en-GB,en-US;q=0.9,en;q=0.8
> 
< HTTP/1.1 204 No Content
< vary: Origin, Access-Control-Request-Headers
< access-control-allow-origin: http://localhost:9876
< access-control-allow-credentials: true
< access-control-allow-methods: GET,HEAD,PUT,PATCH,POST,DELETE
< access-control-allow-headers: authorization
< content-length: 0
< Date: Wed, 29 Jan 2020 16:45:49 GMT
< Connection: keep-alive
< 
* Connection #0 to host 127.0.0.1 left intact
```
Note the header `access-control-allow-headers: authorization`

**Environment (remove any that are not applicable):**
 - Library version: 3.2.4
 - OS: Linux
 - Browser: Any supporting CORS (e.g. Chrome/78.0)

**Additional context**

We were using npm to install this. Somewhat unfortunately, prism-cli does not pin its dependencies' patch versions for the http-server, core, etc. This confused us when we downgraded to prism-cli 3.2.3, since the problem kept occurring: we thought that prism wasn't at fault. It wasn't until we tried on a machine that hadn't upgraded prism that we understood why it was failing.

**Comment by klarose** [comment]
We may also want to update the `vary` header to match what it gave before. Otherwise I could see us getting some weird behaviour due to caching.

**Comment by XVincentX** [comment]
@klarose Yes, apparently we forgot to handle some headers. Can you try this branch out? https://github.com/stoplightio/prism/tree/fix/request-headers

**Comment by vanhoofmaarten** [comment]
Hi everyone. I'm also experiencing CORS errors since upgrading to Prism v3.2.4.

To have an extra set of eyes on this, I've locally setup the changes made in #957. I believe this fix also needs to set an extra _Access-Control-Allow-Origin_ header in the `handler` function. Without it I still receive CORS issues.

I've been able to fix CORS issues locally by simply adding the following piece of code right after the opening brace of the handler function on line 51.

https://github.com/stoplightio/prism/blob/572f4c536e0cdacb54822a852da04b9806c099f8/packages/http-server/src/server.ts#L51

`reply.setHeader('Access-Control-Allow-Origin', request.headers['origin'] || '*');`

I hope this helps. Thanks.

**Comment by XVincentX** [comment]
@vanhoofmaarten Can you try #957 and let me know if that fixes the error?

**Comment by vanhoofmaarten** [comment]
@XVincentX 

#957 currently does fix the issue for the preflight request, but not for the actual request.

`Cross-Origin Request Blocked: The Same Origin Policy disallows reading the remote resource at http://localhost:4010/news/ (Reason: CORS header ‘Access-Control-Allow-Origin’ missing)`

I believe that the code my previous comment can resolve the entire CORS issue.

**Comment by XVincentX** [comment]
@vanhoofmaarten Can you try again now?

**Comment by vanhoofmaarten** [comment]
@XVincentX I can confirm this issue is fixed on my side.

**Comment by XVincentX** [comment]
Nice. This is released in 3.2.5

**Comment by klarose** [comment]
Thanks for the quick turnaround!
